File: task2/C++/task2.cpp
//g++ task2.cpp -o task2.exe
//./task2.exe
#include <iostream>
#include <fstream>
#include <string>
#include <set>

using namespace std;

bool isValidLocalPart(const string& local) {
    // Длина от 6 до 30 символов
    if (local.length() < 6 || local.length() > 30)
        return false;

    // Не может начинаться или заканчиваться точкой
    if (local.front() == '.' || local.back() == '.')
        return false;

    // Запрещены несколько точек подряд
    if (local.find("..") != string::npos)
        return false;

    // Проверка каждого символа
    for (char c : local) {
        bool isAllowed = (c >= 'a' && c <= 'z') ||
                         (c >= '0' && c <= '9') ||
                         c == '.' || c == '+';
        if (!isAllowed)
            return false;
    }
    return true;
}

string normalize(const string& email) {
    size_t atPos = email.find('@');
    if (atPos == string::npos || atPos == 0 || atPos == email.length() - 1)
        return "";

    string local = email.substr(0, atPos);
    string domain = email.substr(atPos);

    if (!isValidLocalPart(local))
        return "";

    string newLocal = "";
    for (char c : local) {
        if (c == '+')
            break;
        if (c != '.')
            newLocal += c;
    }

    return newLocal + domain;
}

int main() {
    set<string> emails;
    int choice;

    cout << "1 - input from console\n2 - input from file\n";
    cin >> choice;

    if (choice == 1) {
        int n;
        cout << "Enter number of emails: ";
        cin >> n;

        string email;
        for (int i = 0; i < n; ++i) {
            cin >> email;
            string norm = normalize(email);
            if (!norm.empty())
                emails.insert(norm);
        }
    } else if (choice == 2) {
        string filename;
        cout << "Enter file name: ";
        cin >> filename;

        ifstream file(filename);
        if (!file.is_open()) {
            cout << "File open error\n";
            return 1;
        }

        string email;
        while (file >> email) {
            string norm = normalize(email);
            if (!norm.empty())
                emails.insert(norm);
        }
        file.close();
    } else {
        cout << "Invalid choice\n";
        return 1;
    }

    cout << "Number of unique emails: " << emails.size() << endl;
    return 0;
}

File: task2/Swift/task2.swift
import Foundation

func isValidLocalPart(_ local: String) -> Bool {
    // Length check
    guard local.count >= 6 && local.count <= 30 else { return false }
    // Cannot start or end with a dot
    guard !local.hasPrefix(".") && !local.hasSuffix(".") else { return false }
    // No consecutive dots
    guard !local.contains("..") else { return false }
    // Allowed characters: a-z, 0-9, ., +
    let allowedSet = CharacterSet.lowercaseLetters
        .union(.decimalDigits)
        .union(CharacterSet(charactersIn: ".+"))
    return local.unicodeScalars.allSatisfy { allowedSet.contains($0) }
}

func normalize(_ email: String) -> String? {
    guard let atPos = email.firstIndex(of: "@") else { return nil }
    let local = String(email[..<atPos])
    let domain = String(email[atPos...])

    guard isValidLocalPart(local) else { return nil }

    var newLocal = ""
    for ch in local {
        if ch == "+" { break }
        if ch != "." { newLocal.append(ch) }
    }
    return newLocal + domain
}

func main() {
    print("1 - input from console\n2 - input from file")
    guard let choiceStr = readLine(), let choice = Int(choiceStr) else {
        print("Invalid input")
        return
    }

    var uniqueEmails = Set<String>()

    if choice == 1 {
        print("Enter number of emails:")
        guard let nStr = readLine(), let n = Int(nStr) else { return }

        for _ in 0..<n {
            guard let email = readLine() else { continue }
            if let norm = normalize(email) {
                uniqueEmails.insert(norm)
            }
        }
    } else if choice == 2 {
        print("Enter file name:")
        guard let filename = readLine() else { return }

        do {
            let content = try String(contentsOfFile: filename, encoding: .utf8)
            let words = content.split { $0.isWhitespace || $0.isNewline }.map(String.init)
            for email in words {
                if let norm = normalize(email) {
                    uniqueEmails.insert(norm)
                }
            }
        } catch {
            print("File open error")
            return
        }
    } else {
        print("Invalid choice")
        return
    }

    print("Number of unique emails:", uniqueEmails.count)
}

main()